DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/niosv/kernels/simple_dma/src/simple_dma.cpp

```diff
@@ -0,0 +1,124 @@
+/* Compile from the /simple_dma_kernel/build directory using the following
+command line
+
+icpx -v -fsycl -fintelfpga -I$INTELFPGAOCLSDKROOT/include ../src/simple_dma.cpp
+-Xshardware -fsycl-link=early -Xstarget=Arria10 -Xsv -o simple_dma.o
+
+  I have to compile this way in order to get more verbose messaging and so that
+interfaces.hpp is found (it's not added to include search path by setvar.sh)
+*/
+
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/interfaces.hpp>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using ext::intel::experimental::property::usm::buffer_location;
+
+static constexpr int BL0 = 0;
+static constexpr int BL1 = 1;
+
+struct SimpleDMA {
+  // declaring the pointers first since they are 64-bit, I want the wide types
+  // first so that the CSRs will be aligned to them
+  register_map_mmhost(
+      BL0,  // buffer space, this needs to be unique per master (don't forget to
+            // encode this value into pointer bits 46:41 otherwise the decoding
+            // might not work as expected)
+      32,   // address width
+      32,  // data width, typically this would be wider but the memory connected
+           // is only 32-bit
+      0,   // latency (choose 0-latency so that waitrequest will work)
+      1,   // 1 = read only
+      15,  // max burst size
+      0,   // aligned transactions only
+      1    // include waitrequest
+      ) unsigned int* const source;
+
+  register_map_mmhost(
+      BL1,  // buffer space, this needs to be unique per master (don't forget to
+            // encode this value into pointer bits 46:41 otherwise the decoding
+            // might not work as expected)
+      32,   // address width
+      32,  // data width, typically this would be wider but the memory connected
+           // is only 32-bit
+      0,   // latency (choose 0-latency so that waitrequest will work)
+      2,   // 2 = write only
+      15,  // max burst size
+      0,   // aligned transactions only
+      1    // include waitrequest
+      ) unsigned int* const destination;
+
+  unsigned int length;  // measured in bytes, must be a multiple of 4
+
+  // Implementation of the DMA kernel.  This accelerator will be controlled via
+  // its Avalon-MM agent interface
+  register_map_interface void operator()() const {
+    // This loop does not handle partial accesses (less than 4 bytes) at the
+    // start and end of the source/destination so ensure they are at least
+    // 4-byte aligned
+    for (unsigned int i = 0; i < (length / 4);
+         i++)  // will be testing with Nios V where an int is 4 bytes
+    {
+      destination[i] = source[i];
+    }
```

Comment:
for users; to explain why we divide legnth by 4. do you think `memcpy` would make more sense here? @jarrodblackburn can you use `memcpy` on the Nios® V softcore processor?

Reply:
ah understood, then I think it should be rewritten as follows:
```suggestion
    // will be testing with Nios V where an int is 4 bytes
    constexpr kSize = length / 4
    for (unsigned int i = 0; i < kSize; i++) {
      destination[i] = source[i];
    }
```

Reply:
memcpy is present for Nios V but this is the kernel code so it's applicable here.  This change is fine given the DMA is only moving integers.  The division was in the original code since I tend to keep everything expressed in bytes (unaligned DMAs get confusing if you deal with a mix of words and bytes).

Reply:
it shouldn't be a constexpr anyway because `length` is a runtime parameter. I'll write it like this:

```c++
  void operator()() const {
    // This loop does not handle partial accesses (less than 4 bytes) at the
    // start and end of the source/destination so ensure they are at least
    // 4-byte aligned
    for (unsigned int i = 0; i < (length_bytes / 4); i++) {
      destination[i] = source[i];
    }
  }
```